File: GeoTimeTracker/app/src/main/java/de/dhbw/geofencinglbs/data/local/AppDatabase.java
package de.dhbw.geofencinglbs.data.local;

import android.content.Context;

import androidx.room.Database;
import androidx.room.Room;
import androidx.room.RoomDatabase;

import de.dhbw.geofencinglbs.model.GeofenceEvent;
import de.dhbw.geofencinglbs.model.GeofenceModel;

/**
 * Room-Datenbankklasse für die gesamte Anwendung.
 * Implementiert das Singleton-Muster.
 */
@Database(entities = {GeofenceModel.class, GeofenceEvent.class}, version = 1, exportSchema = false)
public abstract class AppDatabase extends RoomDatabase {

    private static final String DATABASE_NAME = "geofencing_db";
    private static AppDatabase instance;

    public abstract GeofenceDao geofenceDao();
    public abstract GeofenceEventDao geofenceEventDao();

    public static synchronized AppDatabase getInstance(Context context) {
        if (instance == null) {
            instance = Room.databaseBuilder(
                            context.getApplicationContext(),
                            AppDatabase.class,
                            DATABASE_NAME)
                    .fallbackToDestructiveMigration() // Bei Schema-Änderungen Datenbank neu erstellen
                    .build();
        }
        return instance;
    }
}

File: LBSApp/app/src/main/java/com/example/lbsapp/database/AppDatabase.kt
package com.example.lbsapp.database

import com.example.lbsapp.database.GeofenceDao
import com.example.lbsapp.database.VisitDao
import android.content.Context
import androidx.room.Database
import androidx.room.Room
import androidx.room.RoomDatabase

@Database(
    entities = [GeofenceEntity::class, VisitEntity::class],
    version = 1,
    exportSchema = false // Verhindert die Schema-Export-Warnung
)
abstract class AppDatabase : RoomDatabase() {
    abstract fun geofenceDao(): GeofenceDao
    abstract fun visitDao(): VisitDao

    companion object {
        @Volatile
        private var INSTANCE: AppDatabase? = null

        fun getDatabase(context: Context): AppDatabase {
            return INSTANCE ?: synchronized(this) {
                val instance = Room.databaseBuilder(
                    context.applicationContext,
                    AppDatabase::class.java,
                    "lbs_app_database"
                ).build()
                INSTANCE = instance
                instance
            }
        }
    }
}